### PR TITLE
Adjusts query's CoerceViaIO & RelabelType nodes that are improper for deparsing

### LINKED ISSUE #5033: Varchar Cannot be casted to Int4range with Distributed Tables
```
CREATE TABLE tbl(a INT);

SELECT ((1)::VARCHAR)::int4range FROM tbl;
┌───────────┐
│ int4range │
├───────────┤
└───────────┘
(0 rows)     

SELECT create_distributed_table('tbl', 'a');

SELECT ((1)::VARCHAR)::int4range FROM tbl;
ERROR:  cannot cast type cstring to int4range
CONTEXT:  while executing command on localhost:9701
```

### LINKED ISSUE #5646: Casting varchar to timestamp does not work in a query on a distributed table
While migrating from a standard PostgreSQL DB to one powered by Citus we discovered some prepared statements stopped working with the error message: `ERROR:  cannot cast type cstring to timestamp without time zone`. Same statements work on a normal Postgres instance and they also work agains a local table on Citus. But they do not work on a distributed table. Steps to reproduce:

```sql
CREATE TABLE data
(
    local_time       TIMESTAMP                NOT NULL,
    id               INT4                     NOT NULL,
    value            FLOAT8                   NOT NULL
);

PREPARE test (varchar) AS
    SELECT * FROM data d WHERE d.local_time > cast($1 as timestamp(6) without time zone);
	
EXECUTE test('2008-04-07 00:00:00'); -- works - returns 0 rows	

-- Now do:

SELECT create_distributed_table(table_name := 'data', distribution_column := 'id', shard_count:= 2);

EXECUTE test('2008-04-07 00:00:00'); -- does not work - ERROR:  cannot cast type cstring to timestamp without time zone
```

In the server log we see:

```
2022-01-22 18:47:05.130 UTC [12817] ERROR:  cannot cast type cstring to timestamp without time zone at character 129
2022-01-22 18:47:05.130 UTC [12817] STATEMENT:  SELECT local_time, id, value FROM public.data_102012 d WHERE (local_time OPERATOR(pg_catalog.>) ('2008-04-07 00:00:00'::cstring)::timestamp(6) without time zone)
```

Looks like the query has been rewritten with `::cstring` cast appended. `cstring` does not look correct. 

P.S. just discovered while finishing writing the ticket: seems that the problem has nothing to do with the prepared statement mechanism: issuing the query `SELECT * FROM data d WHERE d.local_time > cast('2008-04-07 00:00:00'::varchar as timestamp(6) without time zone);` does not work when the table is distributed.

Tested on Postgres 13 and 14 with Citus 10.2.

### LINKED ISSUE #6061: Error when casting text to timestamps in prepared statement
Consider the following statements:

```sql
CREATE TABLE test(t timestamptz, user_id text);
SELECT create_distributed_table('test', 'user_id');

PREPARE test_statement(text) AS
SELECT * FROM test WHERE t >= $1::timestamptz;

EXECUTE test_statement('2022-01-01');
```

## Expected result

Query executed successfully 

## Actual result

```
[42846] ERROR: cannot cast type cstring to timestamp with time zone
Where: while executing command on localhost:5432
```

## Environment

Citus version: 11.0-3
Postgres version: 14.4

## Misc

This way of preparing statements is quite common for JDBC applications

### PR DESCRIPTION
**Pull request by naisila**
DESCRIPTION: Adjusts query's CoerceViaIO & RelabelType nodes that are improper for deparsing

The standard planner converts some `::text` casts to `::cstring` and here we convert back because `cstring` is a pseudotype and it cannot be casted to most types. This problem occurs in CoerceViaIO nodes.
There was another problem with RelabelType nodes fixed in the following PR:
https://github.com/citusdata/citus/pull/4580
We undo the changes in that PR, and fix both CoerceViaIO and RelabelType nodes in the planning phase (not in the deparsing phase in ruleutils)

Note to reviewer: I added the test in prepared statements given that this bug was reported twice by users, and both their use-cases involved a prepared statement. Please note that this bug actually has nothing to do with a prepared statement. Feel free to suggest another place where we could put these tests.

For reference, this is a similar PR for a different issue related to parsing https://github.com/citusdata/citus/pull/4580
Thanks @onurctirtir for linking that PR in https://github.com/citusdata/citus/issues/5646#issuecomment-1019823625

Fixes #5646 
Fixes #6061 
Fixes #5033

### CODE REVIEW

File: src/backend/distributed/deparser/ruleutils_13.c [hunk 1]
```diff
@@ -5290,6 +5292,27 @@ get_rule_expr(Node *node, deparse_context *context,
 				CoerceViaIO *iocoerce = (CoerceViaIO *) node;
 				Node	   *arg = (Node *) iocoerce->arg;
 
+				/*
```
Comment (onderkalaci) [inline src/backend/distributed/deparser/ruleutils_13.c:5295]: Why did we end-up doing it on the deparser? I though you were trying to fix this in the planner. There are some disadvantages of doing it in the deparser, here are my thoughts:

1) It is too late :) This is not a deparsing problem, so seems like should be fixed where the problem arise? (e.g., in the planner)
2) Relatively hard to follow, because we don't do these types of modifications during the deparse. Deparse is (should be) a read-only operation.
3) We do not deparse queries that go through standard_planner. This can be confusing addition, the reader of this code can think that we always deparse queries that passed through the standard_planner. There are obvious cases (like complex subqueries) that cannot be parsed back after standard_planner. We only need this because we use some expressions in re-written query, not the whole query.
4) Any change in deparser is a potential conflict with future Postgres versions. I'd rather avoid if there is another way to solve it. I think here we have?

So, overall, I'd be more inclined to see this (or any other change) on the SqlTaskList function instead of here.

Comment (naisila) [reply src/backend/distributed/deparser/ruleutils_13.c:5295]: >I'd rather avoid if there is another way to solve it. I think here we have?

It gets more complicated to figure out where this problematic node is in more complicated queries. The deparsing logic is already visiting all nodes recursively so it does that for us. For example, the following also fails in Citus:
`SELECT cast('2008-04-07 00:00:00'::varchar as timestamp(6) without time zone) AS dummy FROM any_dist_table;`
Not sure whether the list is correct but to give an idea the node could be in
query->jointree->quals
query->targetList
query->returningList
query->jointree->fromlist
...
And for each of these lists we call get_rule_expr recursively to get to the `CoerceViaIO` node.

> Relatively hard to follow, because we don't do these types of modifications during the deparse. Deparse is (should be) a read-only operation.

As I linked here, we already have a very similar addition in this `get_rule_expr` function in Relabel types. Deparse is already not a read-only operation as I stated before, such changes for Citus are very easily done in `get_rule_expr`.

Do you still think we should do this differently?

Comment (onderkalaci) [reply src/backend/distributed/deparser/ruleutils_13.c:5295]: > query->jointree->quals
query->targetList
query->returningList
query->jointree->fromlist
...possibly more lists

All is part of "query" so we should be doing a `query_tree_mutator`, not individual conversions. `query_tree_mutator` traverses the queryTree, see `ResolveExternalParams` as an example.

> As I linked here, we already have a very similar addition in this get_rule_expr function in Relabel types. Deparse is already not a read-only operation as I stated before, such changes for Citus are very easily done in get_rule_expr.

That sounds like not the ideal thing we do. I would NOT expect to call a deparser and let it change the input. We can decide to follow not-good practice here or try to make it more intuitive. 

> Do you still think we should do this differently?

Yes, I do. We should only do this if we had to. Now, I'm even in favor of moving the Relabel type related conversion to the new `query_tree_mutator` we plan to write.

I think the follow up question is this: Do you think is it a good idea to do in the parser? I can see that this is easier, but I'm not very comfortable with such a change. I'd at least want to see a sketch of doing `query_tree_mutator` on the `SQLTaskList` and see how it looks. 

If that turns out to be too complex, we might consider getting back to this solution.

Does that work?

Comment (naisila) [reply src/backend/distributed/deparser/ruleutils_13.c:5295]: > query_tree_mutator traverses the queryTree, see ResolveExternalParams as an example.

I was not aware of this, that simplifies things a lot :)

> Do you think is it a good idea to do in the parser?

If query_tree_mutator works as I am imagining right now it would work, I am all in favor of following your suggestion. The main reason I chose the deparser was that it was doing the traversing work for me.

> Now, I'm even in favor of moving the Relabel type related conversion to the new query_tree_mutator we plan to write.

If all turns out beautifully, +1 on that.

> If that turns out to be too complex, we might consider getting back to this solution. Does that work?

Yes, thanks for your suggestion. I will work on that and update the PR accordingly.

**Comment by naisila** [comment]
@onderkalaci 
> Now, I'm even in favor of moving the Relabel type related conversion to the new query_tree_mutator we plan to write.

I wrote a query tree mutator for modifying both `CoerceIO` and `RelabelType` problematic nodes.
For the `RelabelType` one, I added an extra condition that the collation shouldn't be the default one, because then we are adding extra unneeded "Collate Default" parts.

File: src/backend/distributed/planner/multi_physical_planner.c [hunk 2]
```diff
@@ -5572,3 +5581,65 @@ TaskListHighestTaskId(List *taskList)
...
+
+/*
+ * ModifyProblematicNodes takes an input rewritten query and modifies
+ * nodes which, after going through our planner, pose a problem when
+ * deparsing. So far we have two such type of Nodes that may pose problems:
+ * RelabelType and CoerceIO nodes.
+ * Details will be written in comments in the corresponding if conditions.
+ */
+static Node *
+ModifyProblematicNodes(Node *inputNode)
```
Comment (onderkalaci) [inline src/backend/distributed/planner/multi_physical_planner.c:5594]: `problematicNode` sounds too generic. What about `AdjustImproperNodesForDeparse` or such?

File: src/test/regress/expected/multi_prepare_sql.out [hunk 3]
```diff
@@ -1195,6 +1195,31 @@ EXECUTE countsome; -- no replanning
...
+-- https://github.com/citusdata/citus/issues/5646
+-- https://github.com/citusdata/citus/issues/5033
+CREATE TABLE test(t timestamptz, user_id int);
+SELECT create_distributed_table('test', 'user_id');
+ create_distributed_table
+---------------------------------------------------------------------
+
+(1 row)
+
+PREPARE test_statement(text) AS
```
Comment (onderkalaci) [inline src/test/regress/expected/multi_prepare_sql.out:1210]: I'd put this test to arbitrary configs tests, which should cover a lot more tests. Can we move these to https://github.com/citusdata/citus/blob/main/src/test/regress/sql/prepared_statements_4.sql or such?

Please remember that you should create the tables in `prepared_statements_create_load ` to use in prepared_statements_4``

File: src/test/regress/expected/multi_prepare_sql.out [hunk 4]
```diff
@@ -1195,6 +1195,31 @@ EXECUTE countsome; -- no replanning
...
+SELECT user_id FROM test WHERE t >= $1::timestamptz ORDER BY user_id;
```
Comment (onderkalaci) [inline src/test/regress/expected/multi_prepare_sql.out:1211]: Lets expand for different planners:
```SQL
PREPARE test_statement_regular(text) AS
SELECT user_id FROM test WHERE t >= $1::timestamptz ORDER BY user_id;


PREPARE test_statement_router(int, text) AS
SELECT user_id FROM test WHERE user_id = $1 AND t >= $2::timestamptz ORDER BY user_id;

PREPARE test_statement_repartition(int, text) AS
SELECT count(*) FROM test t1 JOIN test t2 USING (t) WHERE t1.user_id = $1 AND t >= $2::timestamptz;


PREPARE test_statement_cte(text, text) AS
with cte_1 as (SELECT user_id, t FROM test WHERE t >= $1::timestamptz ORDER BY user_id LIMIT 1) SELECT * FROM cte_1 WHERE t <=$2::timestamp ;


PREPARE test_statement_insert(int, text) AS
INSERT INTO test VALUES ($2::timestamptz, $1);
```

File: src/test/regress/expected/multi_prepare_sql.out [hunk 5]
```diff
@@ -1195,6 +1195,31 @@ EXECUTE countsome; -- no replanning
...
+INSERT INTO test VALUES ('2022-02-02', 0);
+INSERT INTO test VALUES ('2022-01-01', 1);
+INSERT INTO test VALUES ('2021-01-01', 2);
+EXECUTE test_statement('2022-01-01');
```
Comment (onderkalaci) [inline src/test/regress/expected/multi_prepare_sql.out:1215]: Executing prepared statements >=6 times is always useful (or even required) as Postgres does certain changes on the query after that

File: src/backend/distributed/planner/multi_physical_planner.c [hunk 6]
```diff
@@ -5572,3 +5581,65 @@ TaskListHighestTaskId(List *taskList)
...
+		Node *arg = (Node *) iocoerce->arg;
+
+		if (IsA(arg, Const) && ((Const *) arg)->consttype == CSTRINGOID)
+		{
+			Const *cstringToText = (Const *) arg;
+
+			cstringToText->consttype = TEXTOID;
+			cstringToText->constlen = -1;
+
+			Type textType = typeidType(TEXTOID);
```
Comment (onderkalaci) [inline src/backend/distributed/planner/multi_physical_planner.c:5629]: what about NULL values? PG seems to do something different:

```C
		/*
		 * We assume here that UNKNOWN's internal representation is the same
		 * as CSTRING.
		 */
		if (!con->constisnull)
			newcon->constvalue = stringTypeDatum(baseType,
												 DatumGetCString(con->constvalue),
												 inputTypeMod);
		else
			newcon->constvalue = stringTypeDatum(baseType,
												 NULL,
												 inputTypeMod);
```

I think some of our tests should trigger NULL values (also good to check for relable types for NULL values? Would we crash?

Comment (onderkalaci) [reply src/backend/distributed/planner/multi_physical_planner.c:5629]: Or, maybe should we use makeConst instead of stringTypeDatum?

File: src/test/regress/expected/multi_prepare_sql.out (hunk 4 above)
Comment (naisila) [reply src/test/regress/expected/multi_prepare_sql.out:1211]: Good idea, thanks for providing the prepared statements.

File: src/backend/distributed/planner/multi_physical_planner.c (hunk 6 above)
Comment (naisila) [reply src/backend/distributed/planner/multi_physical_planner.c:5629]: > Or, maybe should we use makeConst instead of stringTypeDatum?

`makeConst` already assumes we are passing a correct Datum for the text type, so we need to use `stringTypeDatum` to obtain a correct Datum for text type.

Comment (naisila) [reply src/backend/distributed/planner/multi_physical_planner.c:5629]: > also good to check for relable types for NULL values

If arg is NULL it will be handled from the CollateExpr node since we just pass that arg. 
`collateExpr->arg = relabelType->arg;`
This if statement already keeps us safe from any other NULL/invalidOid cases:
```C
if (IsA(inputNode, RelabelType) &&
    OidIsValid(((RelabelType *) inputNode)->resultcollid) &&
    ((RelabelType *) inputNode)->resultcollid != DEFAULT_COLLATION_OID)
```

**Comment by naisila** [comment]
Note: Not ready for review yet - `test_statement_cte` is not supposed to return empty set.

**Comment by naisila** [comment]
@onderkalaci turns out that even in my local using == with timestamptz is tricky: in fact in my local even more configurations fail.

I used `timestamp` instead since it serves the exact same purpose for these tests.

Thanks

File: src/backend/distributed/planner/multi_physical_planner.c [hunk 7]
```diff
@@ -2691,6 +2697,9 @@ SqlTaskList(Job *job)
 	List *fragmentCombinationList = FragmentCombinationList(rangeTableFragmentsList,
 															jobQuery, dependentJobList);
 
+	/* Adjust RelabelType and CoerceViaIO nodes that are improper for deparsing */
+	jobQuery = (Query *) AdjustImproperForDeparseNodes((Node *) jobQuery);
```
Comment (marcocitus) [inline src/backend/distributed/planner/multi_physical_planner.c:2701]: note that query_tree_mutator makes a deep copy of the query tree, so the overhead of always doing this can be significant for complex queries

I assume that's why RelabelTypeToCollateExpr is so far used in the deparser

Comment (naisila) [reply src/backend/distributed/planner/multi_physical_planner.c:2701]: I have added `QTW_DONT_COPY_QUERY` flag to the `query_tree_mutator` function such that it's modified in place

```C
return (Node *) query_tree_mutator((Query *) inputNode, AdjustImproperForDeparseNodes,
				   NULL, QTW_DONT_COPY_QUERY);
```

From Postgres:
```C
/*
 * query_tree_mutator --- initiate modification of a Query's expressions
 *
...
 * Normally the top-level Query node itself is copied, but some callers want
 * it to be modified in-place; they must pass QTW_DONT_COPY_QUERY in flags.
 * All modified substructure is safely copied in any case.
 */
Query *
query_tree_mutator(Query *query,
	           Node *(*mutator) (),
		   void *context,
		   int flags)
```

Comment (naisila) [reply src/backend/distributed/planner/multi_physical_planner.c:2701]: Oops I missed `All modified substructure is safely copied in any case.`

Comment (marcocitus) [reply src/backend/distributed/planner/multi_physical_planner.c:2701]: Well, that only takes care of the top-level Query struct and all the expression nodes below it are still copied.

Given the complexity of updating ruleutils, making the fix separately kind of makes sense, but it would be good to add a check (a walker that checks whether any relabel types exist).

File: src/backend/distributed/planner/multi_physical_planner.c [hunk 8]
```diff
@@ -2691,6 +2697,12 @@ SqlTaskList(Job *job)
 	List *fragmentCombinationList = FragmentCombinationList(rangeTableFragmentsList,
 															jobQuery, dependentJobList);
 
+	/* Adjust RelabelType and CoerceViaIO nodes that are improper for deparsing */
+	if (QueryTreeHasRelabelTypeOrCoerceViaIONodes((Node *) jobQuery))
+	{
```
Comment (onderkalaci) [inline src/backend/distributed/planner/multi_physical_planner.c:2702]: would be nice to have a comment on why we first check the query tree and then only do AdjustImproperForDeparseNodes later. It might not be obvious for the readers.

**Review by onderkalaci — APPROVED** [review]
Thanks for the fix and applying @marcocitus's suggestion. 

I'm happy that we fixed this on the planner not deparser. Let's wait before merging and give a chance for @marcocitus to have a look, just in case he has additional comments

**Comment by onderkalaci** [comment]
@naisila is there a reason for not merging this?

**Comment by onderkalaci** [comment]
> @naisila is there a reason for not merging this?

I mean, lets not get this to 11.2 at this point, but right after the release, we can get to 11.3 (if there is not something I'm missing)

**Comment by naisila** [comment]
> @naisila is there a reason for not merging this?

I was a bit worried about query performance. I wonder, should I run a benchmark for this? The extra query tree walker whenever we have a distributed query is usually not a costly operation, but is it negligible enough even for complex queries? I wasn't sure.

**Comment by naisila** [comment]
I tested the performance of a "complex" query on my local machine. Query has 9 subqueries. Avg query time is **30ms** in both cases.
#### Setup:
```SQL
CREATE TABLE test(t timestamp, user_id int);
SELECT create_distributed_table('test', 'user_id');
INSERT INTO test VALUES ('2022-02-02', 0);
INSERT INTO test VALUES ('2022-01-01', 1);
INSERT INTO test VALUES ('2021-01-01', 2);
```
#### Main branch:
```SQL
SELECT * FROM (SELECT * FROM (SELECT * FROM (SELECT * FROM (SELECT * FROM (SELECT * FROM (SELECT * FROM
(SELECT * FROM (SELECT * FROM (SELECT * FROM (SELECT * FROM test WHERE t >= '2022-01-01'::timestamp) t1
WHERE t >= '2022-01-01'::timestamp) t2 WHERE t >= '2022-01-01'::timestamp) t3 WHERE t >= '2022-01-01'::timestamp) t4
WHERE t >= '2022-01-01'::timestamp) t5 WHERE t >= '2022-01-01'::timestamp) t6 WHERE t >= '2022-01-01'::timestamp) t7 
WHERE t >= '2022-01-01'::timestamp) t8 WHERE t >= '2022-01-01'::timestamp) t9 WHERE t >= '2022-01-01'::timestamp) t10 
WHERE t >= '2022-01-01'::timestamp;
```

#### This branch
```SQL
SELECT * FROM (SELECT * FROM (SELECT * FROM (SELECT * FROM (SELECT * FROM (SELECT * FROM (SELECT * FROM
(SELECT * FROM (SELECT * FROM (SELECT * FROM (SELECT * FROM test WHERE t >= '2022-01-01'::text::timestamp) t1
WHERE t >= '2022-01-01'::text::timestamp) t2 WHERE t >= '2022-01-01'::text::timestamp) t3 WHERE t >= '2022-01-01'::text::timestamp) t4
WHERE t >= '2022-01-01'::text::timestamp) t5 WHERE t >= '2022-01-01'::text::timestamp) t6 WHERE t >= '2022-01-01'::text::timestamp) t7
WHERE t >= '2022-01-01'::timestamp) t8 WHERE t >= '2022-01-01'::text::timestamp) t9 WHERE t >= '2022-01-01'::text::timestamp) t10
WHERE t >= '2022-01-01'::text::timestamp;
```
